source/crud/crud.md

```diff
@@ -815,6 +815,12 @@ database-level aggregation will allow users to receive a cursor from these colle
 
 ##### Insert, Update, Replace, Delete, and Bulk Writes
 
+###### Generated identifiers
+
+The insert and bulk insert operations described below MUST generate identifiers for all documents that do not already
+have them. These identifiers SHOULD be prepended to the document so they are the first field, in order to prevent the
+server from spending time re-ordering the document.
```

Comment:
It is implied, but may be we should make it abundantly clear, a driver should _not_ re-order an _id if it is already present right?

I would say yes, but also not sure if it matters and it may be good to do. If the goal is performance then we should weigh re-ordering vs the cost of leaving the _id out of place 🤔

Reply:
The cost of re-ordering on the client-side is language-dependent. I'm not well-versed in how performant the server's re-ordering operation is, but it very likely could be faster than many drivers.

Reply:
With respect to DRIVERS-1408, I don't see any reason to add re-ordering logic for user-supplied documents that already contain an `_id`. Just ensure that if the driver is generating an ID, the field gets prepended.

Reply:
I agree. I would take it further and say that drivers SHOULD NOT re-order given documents if the user has given a structure that has order. The choice to place the _id in the sub/optimal position along with any other positioning choices should be preserved. (TBH, mostly because it is useful for testing)

Maybe it could be a MUST NOT too 🤔 as long as you've been given an ordered object, no intervention should be done.

In JS the `_id` is always appended because JS objects are ordered by key insertion order so there is no way to put the _id in front. Naively we could special-case _id in our BSON library but that would indiscriminately reorder that key no matter where it is used. I am thinking ahead for my own driver here, we want a solution that _only_ fixes this for driver generated `_id` fields.

Reply:
I'd leave it up to each driver on how to handle already-existing `_id` fields. The Python driver reorders user-supplied `_id` keys during the BSON conversion, but other drivers may be unable or unwilling to do the same.

Reply:
That is fair and I wouldn't want to imply python should change that, nor do I want us to imply to drivers that do not re-order user `_id`s should do so. So I think it is fair to leave open to each language but we should capture this conversation in some form. 

- Generated _ids SHOULD be put first. 
- User _ids MAY be left as is. (hopefully that is enough to imply, "MAY be re-ordered as well"?)